File: roooom/app/src/main/java/com/example/roooom/MainActivity.java
package com.example.roooom;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.example.roooom.AccessDataBase.DataManager;
import com.example.roooom.AccessDataBase.RxForData;

public class MainActivity extends AppCompatActivity implements DataManager {

    DataManager dataManager;
    RxForData rx;
    String email;
    String pass;

    EditText userEmail;
    EditText userPass;


    Button sendData;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initView();

        rx = new RxForData(this);
        sendData.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {

                //Strings Form EditText:
                email = userEmail.getText().toString();
                pass = userPass.getText().toString();

                if(email.isEmpty() || email == null || pass.isEmpty() || pass == null){
                    onFailure();
                }else{
                    if (email.length() < 9 || pass.length() < 9 ){

                        onFailure();
                    }else {
                        onSuccess();
                        rx.addToDb(dataManager ,email , pass);
                        rx.getAllData();
                    }
                }
            }
        });

    }

    private void initView() {

        sendData = findViewById(R.id.submit);
        userEmail = findViewById(R.id.email);
        userPass = findViewById(R.id.pass);



    }

    @Override
    public void onSuccess() {
        makeToast("Data Added Successfully");
    }

    @Override
    public void onFailure() {
        makeToast("Failed To Add Data To DataBase");
    }

    @Override
    public void onUpdate() {
        makeToast("Data Updated");
    }


    public void makeToast(String message){
        Toast.makeText(this , message , Toast.LENGTH_SHORT).show();
    }

}

File: KotlinCoroutines/app/src/main/java/com/example/kotlin_coroutines/MainActivity.kt
package com.example.kotlin_coroutines

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button
import android.widget.TextView
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.Dispatchers.Main
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class MainActivity : AppCompatActivity() {

    lateinit var result : TextView
    lateinit var button : Button


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        initView()

        button.setOnClickListener({
                CoroutineScope(IO).launch {
                    getFirst()
                }
            }
        )
    }

    private suspend fun getFirst(){
        val result1 = firstResult()
        setToTextView(result1)

        val result2 = secondResult()
        setToTextView(result2)
    }

    private suspend fun setToTextView(result1: String) {
        withContext(Main){
            result.text = "${result.text} -- $result1 \n"
        }
    }

    private suspend fun getSecond(){
        val result2 = secondResult()
        println("Result : $result2")
    }

    private suspend fun firstResult() : String{
        delay(2000)
        return "#First Result"
    }

    private suspend fun secondResult() : String{
        delay(2000)
        return "#Second Result"
    }

    private fun initView() {
        button = findViewById(R.id.button)
        result = findViewById(R.id.resultView)
    }
}